File: src/main/scala/spark/OnlineDomain.scala
package spark

import com.alibaba.fastjson.{JSON, JSONObject}
import com.mongodb.spark.MongoSpark
import org.apache.spark.{SparkConf, SparkContext}
import org.bson.Document
import util.StringUtil

object OnlineDomain {
  def main(args: Array[String]): Unit = {

    val domainInclude = Array("VR", "WorldRecords", "IDIOM", "POEM", "JOKE", "REMINDER", "HISTORY", "HTWHYS", "DISEASE", "STORY")

    val conf = new SparkConf().set("spark.mongodb.output.uri", "mongodb://10.66.188.17:27017/semantic.semantic_tencent_online_domain")
    val sc = new SparkContext(conf)

    val input = sc.textFile("hdfs://hadoop1:9000"+args(0))

    val result = input.map(record => JSON.parseObject(record))
      .filter(record => {
        val domain = record.getString("return_domain")
        var flag = true
        try{
          if(record.getJSONObject("query").getJSONObject("result").getInteger("sourceFlag") != 5){
            flag = false
          }
        } catch {
          case _: Exception => flag = false
        }
        flag &&
        StringUtil.isNotEmpty(domain) &&
          domainInclude.contains(domain) &&
          StringUtil.isNotEmpty(record.getString("query_text"))
      })
      .map(record => {
        val value = new JSONObject()
        value.put("domain", record.getString("return_domain"))
        value.put("intent", record.getString("return_intent"))
        value.put("semantic", record.getJSONObject("return_semantic"))
        value.put("sourceFlag", record.getJSONObject("query").getJSONObject("result").getInteger("sourceFlag"))
        (record.getString("query_text"), value.toString())
      })
      .groupByKey()
      .map(record => {
        val value = JSON.parseObject(record._2.iterator.next())
        val total = record._2.iterator.size
        new Document()
          .append("query_text", record._1)
          .append("domain", value.getString("domain"))
          .append("intent", value.getString("intent"))
          .append("semantic", value.getJSONObject("semantic"))
          .append("sourceFlag", value.getInteger("sourceFlag"))
          .append("count", total)
      })

    MongoSpark.save(result)

    sc.stop()

  }
}

File: src/main/java/spark/OnlineDomain.java
package spark;

import com.alibaba.fastjson.JSONObject;
import com.google.common.collect.Iterables;
import com.mongodb.spark.MongoSpark;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.bson.Document;
import scala.Tuple2;
import util.StringUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * 924上线领域，每15天导一次
 */

public class OnlineDomain {

    private static final List<String> DOMAIN = new ArrayList<String>(){
        {
            add("MUSIC");
            add("POEM");
            add("JOKE");
            add("WorldRecords");
            add("IDIOM");
            add("REMINDER");
            add("STORY");
            add("PhoneCall");
            add("FOOD");
            add("WEATHER");
        }
    };

    public static void main(String[] args) {

        SparkConf conf = new SparkConf()
                .set("spark.mongodb.output.uri", "mongodb://10.66.188.17:27017/semantic.semantic_924");
        JavaSparkContext jsc = new JavaSparkContext(conf);

        JavaRDD<String> input = jsc.textFile("hdfs://hadoop1:9000"+args[0]);

        JavaRDD<Document> result = input
                .map(JSONObject::parseObject)
                .filter(record -> {
                    String domain = record.getString("return_domain");
                    return StringUtil.isNotEmpty(domain) &&
                            DOMAIN.contains(domain) &&
                            StringUtil.isNotEmpty("query_text");
                })
                .mapToPair(record -> {
                    JSONObject value = new JSONObject();
                    value.put("domain", record.getString("return_domain"));
                    value.put("intent", record.getString("return_intent"));
                    value.put("semantic", record.getJSONObject("return_semantic"));
                    return new Tuple2<>(record.getString("query_text"), value.toString());
                })
                .groupByKey()
                .map(record -> {
                    int total = Iterables.size(record._2);
                    JSONObject value = JSONObject.parseObject(record._2.iterator().next());
                    Document doc = new Document();
                    doc.put("query_text", record._1);
                    doc.put("domain", value.getString("domain"));
                    doc.put("intent", value.getString("intent"));
                    doc.put("semantic", value.getJSONObject("semantic"));
                    doc.put("count", total);
                    return doc;
                });

        MongoSpark.save(result);

        jsc.stop();
    }

}
